### ISSUE TITLE
How to access reward/action/state during training?

### ISSUE THREAD

**Issue by Zebin-Li** [takuseno/d3rlpy#306]
Hi,

Thank you very much for the wonderful work! I have a quick requirement that I want to check the reward/action/state, etc., during each step of the training when using offline dataset. However, the .fit() function encloses everything so how can I access those? Thank you very much!

Best,
Zebin

**Comment by Zebin-Li** [comment]
I don't have an environment only the offline dataset, so the code like this 
`env_evaluator = EnvironmentEvaluator(env)`
`rewards = env_evaluator(dqn, dataset=None)`
does not work I think.

**Comment by takuseno** [comment]
@Zebin-Li Hi, thank for the issue. I actually don't understand your request here. What do you want to do?

**Comment by Zebin-Li** [comment]
@takuseno Thank you for your quick response! 
Let me rephrase my question. In general RL training process, we have something like: 
```
for _ in range(100):
    states = environment.reset()
    terminal = False
    while not terminal:
        actions = agent.act(states=states)
        states, terminal, reward = environment.execute(actions=actions)
        agent.observe(terminal=terminal, reward=reward)
```
So that I can access to the state/reward/action in each training step to check them or perform some customized operations.
However, in d3rlpy, the training process is done by the algo.fit() function, where I cannot explicitly get the state/action/reward in each training step. So should I do it?

Besides, I found that the reward can be accessed by
```
env_evaluator = EnvironmentEvaluator(env)
rewards = env_evaluator(dqn, dataset=None)
```
But I am using a customized offline dataset and I do not have the "env", so I guess it does not work for me. Hope this time my question is more clear! Thank you!

**Comment by Zebin-Li** [comment]
Hi,

I am trying to write a callback function and trying to access to the reward at each step during training by something like
```
for episode in episodes:
        for i in range(episode.transition_count):
            transition = transition_picker(episode, i)
            rewards.append(transition.reward)
```
However, the transition is a protected attribute, so how should I modify to access the transition from outside? Thank you!

**Comment by takuseno** [comment]
Sorry for the late response. To be honest, I still don't understand what you want to achieve. Do you want to modify transion when it's sampled?

**Comment by Zebin-Li** [comment]
Thank you for the quick response! Actually what I want to do is to check the "reward" during the training process. In RL training, when we take an action, transit to a new state and receive a reward. Now I want to explicitly monitor the reward the agent received at each timestep during the training process. Hope this helps!

**Comment by Zebin-Li** [comment]
"Modifying transition when it's sampled" is something close to what I am trying to, since if I can modify the transition, then I can check the reward as well.

**Comment by takuseno** [comment]
Sorry, I actually don't understand why you want to monitor rewards during training since you already have offline datasets. You can simply check those:
```
for episode in episodes:
    print(episode.rewards)
```

**Comment by Zebin-Li** [comment]
Thank you for your quick response! Actually I want to plot the reward curve over steps or cumulative reward curve over episodes to evaluate or check whether the trained policy is good. I found this issue https://github.com/takuseno/d3rlpy/issues/218 had the similar content which asking about plotting cumulative reward on Tensorboard. I notice that there is a way to do so by using `evaluate_on_environment(env)`, however, I do not have such an RL environment (i.e., `env`) since my task is a real-world application and I want to do the evaluation still on the offline dataset. Hope this make me more clear!

**Comment by takuseno** [comment]
Hmm, probably, you seem to misunderstand offline RL...? Monitoring reward values fed to algorithms during training is completely different from environmental returns during evaluation. So, it's no good idea to do it. Basically, in offline deep RL, there is no solid way to evaluate policies during training. Please check publications about offline policy evaluation (OPE).

**Comment by Zebin-Li** [comment]
Thank you for the response! Please see the plot below which is from the AWAC paper. Actually I want to plot a similar reward/return curve. 
![image](https://github.com/takuseno/d3rlpy/assets/52910998/f95ad399-c5c8-4f05-af28-52d1c0f56054)

**Comment by Zebin-Li** [comment]
I know there is an environment required, but for my task, it is a real-world application so it is hard to create such an environment and this is why I raise this issue to see what I can do to modify to suit my needs, thank you!

**Comment by takuseno** [comment]
As I mentioned above, the best thing I can suggest here is to check publications about offline policy evaluation (OPE). d3rlpy supports one OPE method described here:
https://d3rlpy.readthedocs.io/en/v2.0.4/references/off_policy_evaluation.html
The method is designed based on this paper:
- https://arxiv.org/abs/2007.09055

I recommend you start reading the paper as the first step.

**Comment by takuseno** [comment]
Another thing I can suggest here is to redirect you to this repository:
- https://github.com/hakuhodo-technologies/scope-rl

This repository highly depends on d3rlpy as a core RL component and they build offline policy evaluation pipelines on top of it. It might be worth checking it.

**Comment by Zebin-Li** [comment]
Thank you for the resource! I will check them.

**Comment by takuseno** [comment]
Let me close this issue. Feel free to reopen this if you have further discussion.
